### PR TITLE
Support aborting unloading

### PR DESCRIPTION
**Pull request by Ashod**
- leaflet: explanatory message when loading fails while unloading
- wsd: allow new views while unloading
- wsd: log discovery.xml response at trace level

### CODE REVIEW

File: wsd/DocumentBroker.cpp [hunk 1]
```diff
@@ -1975,6 +1980,12 @@ size_t DocumentBroker::removeSession(const std::string& id)
         // If last editable, save and don't remove until after uploading to storage.
         if (!lastEditableSession || !autoSave(isPossiblyModified(), dontSaveIfUnmodified))
```
Comment (mmeeks) [inline wsd/DocumentBroker.cpp:1981]: I worry about the last readonly session - how do we mark to destroy for those sessions ?

File: wsd/DocumentBroker.cpp [hunk 2]
```diff
@@ -1934,10 +1941,8 @@ size_t DocumentBroker::removeSession(const std::string& id)
         }
         std::shared_ptr<ClientSession> session = it->second;
 
-        // Last view going away, can destroy.
-        if (_sessions.size() <= 1)
```
Comment (mmeeks) [inline wsd/DocumentBroker.cpp:1938]: Ensuring we handle _sessions.size == 0 carefully is important to not leak sessions of course.

File: wsd/DocumentBroker.cpp [hunk 3]
```diff
@@ -1999,6 +2010,14 @@ void DocumentBroker::disconnectSessionInternal(const std::string& id)
             LOOLWSD::dumpEndSessionTrace(getJailId(), id, _uriOrig);
 #endif
 
+            if (_docState.isUnloadRequested())
+            {
+                // We must be the last session, flag to destroy if unload is requested.
+                assert(_sessions.size() == 1 && "Unload-requested with multiple sessions.");
+                _docState.markToDestroy();
+                LOG_TRC("Unload requested while disconnecting session [" << id << "], marking to destroy.");
```
Comment (mmeeks) [inline wsd/DocumentBroker.cpp:2018]: lets have _sessions.size() in that trace - just in case.

File: wsd/DocumentBroker.cpp [hunk 4]
```diff
@@ -3066,6 +3088,7 @@ void DocumentBroker::dumpState(std::ostream& os)
     os << "\n  doc activity: " << DocumentState::toString(_docState.activity());
     if (_docState.activity() == DocumentState::Activity::Rename)
         os << "\n  (new name: " << _renameFilename << ')';
+    os << "\n  unload requested: " << _docState.isUnloadRequested();
```
Comment (mmeeks) [inline wsd/DocumentBroker.cpp:3091]: glad to see it there =)

**Review by mmeeks — CHANGES_REQUESTED** [review]
I would like every use-case of: isMarkedToDestroy properly audited.

Here is the first one I saw that looked dodgy:

DocumentBroker::NeedToUpload DocumentBroker::needToUploadToStorage() const
{
    // When destroying, we might have to force uploading if always_save_on_exit=true.
    if (isMarkedToDestroy())

I didn't read further. I assume we have a unit test for that though (?).

In general I'd prefer a 'state' that moves from Normal -> Unloading -> ToDestroy rather than two booleans but ... perhaps we can cope with that =)

File: wsd/DocumentBroker.cpp (hunk 1 above)
Comment (Ashod) [reply wsd/DocumentBroker.cpp:1981]: Good point. Improved.

File: wsd/DocumentBroker.cpp (hunk 2 above)
Comment (Ashod) [reply wsd/DocumentBroker.cpp:1938]: Restored.

Comment (kendy) [reply wsd/DocumentBroker.cpp:1944]: Not sure about the previous version of the patch; the current one turned <= 1 into == 1; but as we talked with @Ashod , we cannot get this far with empty _sessions (we search the id in there, and return if it was not found).

**Comment by kendy** [comment]
Tested locally, and it does what it is supposed to do for me :-)  Also checked that the number of sessions is behaving as it should (eg. we temporarily have 2 sessions when the browser window is closed, but another browser window connects to the document while autosaving due to the closed window).

Tried to review the code too; I think we are good, thank you!

**Pull request merged**